### ISSUE TITLE
Recompilation of purely static `Modules`

### ISSUE THREAD

**Issue by KeAWang** [patrick-kidger/equinox#291]
Consider the following code:
```python
import jax
import jax.numpy as jnp
import equinox as eqx
from functools import partial

# define a eqx.Module that is hashable (since it has no unhashable leaves)
class Model(eqx.Module):
    def f(self):
        return
    
model = Model()
# We get a different function (Partial function) each time we __get__ a bound method
f1 = model.f
f2 = model.f
assert not (f1 is f2)
print(hash(f1))
print(hash(f2))

# Let's try using the bound method as a static argument; now jax.jit cares only about __hash__ and __eq__ of the static argument
@partial(jax.jit, static_argnums=0)
def apply(f):
    return f()

# This causes a recompilation since the hash of f1 and f2 are different
with jax.log_compiles(True):
    apply(f1)
    apply(f2)
```
tl;dr bound methods have a different hash each time they are accessed because of this wrapper: https://github.com/patrick-kidger/equinox/blob/2cf6fecb1fb275a13d4931ec79fa91e4f38cc315/equinox/module.py#L64. We should probably memoize the result of `__get__` so that we don't get a new `Partial` every time we access a bound method (assuming the instance of `Module` is immutable).

**Comment by patrick-kidger** [comment]
So each call to `__get__` does produce results with matching pytree structures. As such: try passing `f1` and `f2` as traced, rather than static, arguments.

(Also take a look at `eqx.filter_jit`, which does away with the whole traced-vs-static thing altogether. It turns out that you can always infer that from the argument types.)

**Comment by KeAWang** [comment]
I'm wondering if it still makes sense to also cache the `Partial` just in case people use `jax.jit` instead of `filter_jit`? It's also a bit surprising that you get a different object each time you access a bound method.

For example, we could add something like
```python
def __get__(self, instance, owner):
    if instance is None:
        return self.method
    try:
        wrapped_method = getattr(self, "wrapped_method")
    except AttributeError:
        wrapped_method = jtu.Partial(self.method, instance)
        self.wrapped_method = wrapped_method
    return wrapped_method
```

**Comment by patrick-kidger** [comment]
So we definitely couldn't *cache* the partial: it's not possible to mutate `self` by doing `self.wrapped_method = wrapped_method`, as modules are immutable.

But what we could do is write our own implementation of `jtu.Partial`, that has a stable hash. Maybe something like:
```python
class _wrap_method:
    def __init__(self, method):
        self.method = method
        ft.update_wrapper(self, method)  # taking the time to fix up `_wrap_method`'s wrapping whilst I'm here.

    def __get__(self, instance, owner):
        if instance is None:
            return self.method
        return _BoundMethod(self.method, instance)

class _BoundMethod(Module):
    func: types.FunctionType = static_field()
    self: Module

    def __call__(self, *args, **kwargs):
        return self.func(self.self, *args, **kwargs)
```
